nexus/src/app/mod.rs

```diff
@@ -153,6 +153,24 @@ impl Nexus {
         let db_datastore = Arc::new(db::DataStore::new(Arc::clone(&pool)));
         db_datastore.register_producers(&producer_registry);
 
+        // Keep looping until we find that the schema matches our expectation.
+        const EXPECTED_VERSION: SemverVersion = SemverVersion::new(1, 0, 0);
+        loop {
+            match db_datastore.database_schema_version().await {
+                Ok(version) => {
+                    if version == nexus_db_model::schema::SCHEMA_VERSION {
+                        break;
+                    }
+                    let observed = version.0;
+                    warn!(log, "Incompatible database schema: Saw {observed}, expected {EXPECTED_VERSION}");
+                }
+                Err(e) => {
+                    warn!(log, "Cannot read database schema version: {e}");
+                }
+            }
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
```

Comment:
I'd suggest sleeping longer.  Ideally I think we'd use backoff if we're failing to read the schema at all, and a fixed sleep (like 60 seconds?) if we successfully read the wrong schema.

(The thing I'm most worried about here is if we're recovering from a database outage and hitting the database too hard.  Secondarily it seems unnecessary to poll every second for something that's unlikely to change that soon.)

Reply:
WDYT about using backoff? I have a test for this behavior, so I kinda want to avoid using a 60-second timeout, unless you think I should plumb this through as a configuration parameter?

Reply:
Yeah let's just do backoff for now.

The thing that makes me a little uneasy about it is that our max backoff used to be an hour, which I still think is more appropriate for overload conditions than the current max of about 3 minutes.  (I understood that we changed it to 3 minutes basically because #3082 was too painful otherwise.)  At some point we're going to need to bump that up a lot more again.  And an hour is a long time to wait in this particular case.  (This condition also doesn't really have the backoff property that the client's own behavior might be affecting its ability to succeed, so backoff is a little weird.)

But in practice I think this will be fine for the foreseeable future so let's go with it.